chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+//     to the current chain head.
+//
+// This function is crucial for maintaining index integrity, especially after chain reorgs.
+// It ensures that the index accurately reflects the current state of the blockchain.
+func (si *SqliteIndexer) ReconcileWithChain(ctx context.Context, currHead *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	if currHead == nil {
+		return nil
+	}
+
+	return withTx(ctx, si.db, func(tx *sql.Tx) error {
+		row := tx.StmtContext(ctx, si.countMessagesStmt).QueryRowContext(ctx)
+		var result int64
+		if err := row.Scan(&result); err != nil {
+			return xerrors.Errorf("error counting messages: %w", err)
+		}
+		if result == 0 {
+			return nil
+		}
+
+		// Find the minimum applied tipset in the index; this will mark the end of the reconciliation walk
+		row = tx.StmtContext(ctx, si.minNonRevertedHeightStmt).QueryRowContext(ctx)
+		if err := row.Scan(&result); err != nil {
+			return xerrors.Errorf("error finding boundary epoch: %w", err)
+		}
+
+		boundaryEpoch := abi.ChainEpoch(result)
+
+		var tipsetStack []*types.TipSet
+
+		curTs := currHead
+		log.Infof("Starting chain reconciliation from height %d", currHead.Height())
+		for curTs != nil && curTs.Height() >= boundaryEpoch {
+			tsKeyCidBytes, err := toTipsetKeyCidBytes(curTs)
+			if err != nil {
+				return xerrors.Errorf("error computing tipset cid: %w", err)
+			}
+
+			var exists bool
+			err = tx.StmtContext(ctx, si.tipsetExistsNotRevertedStmt).QueryRowContext(ctx, tsKeyCidBytes).Scan(&exists)
+			if err != nil {
+				return xerrors.Errorf("error checking if tipset exists and is not reverted: %w", err)
+			}
+
+			if exists {
+				// found it!
+				boundaryEpoch = curTs.Height() + 1
+				log.Infof("Found matching tipset at height %d, setting boundary epoch to %d", curTs.Height(), boundaryEpoch)
+				break
+			}
+			tipsetStack = append(tipsetStack, curTs)
+
+			// walk up
+			parents := curTs.Parents()
+			curTs, err = si.cs.GetTipSetFromKey(ctx, parents)
+			if err != nil {
+				return xerrors.Errorf("error walking chain: %w", err)
+			}
+		}
+
+		if curTs == nil {
+			log.Warn("ReconcileWithChain reached genesis without finding matching tipset")
+		}
+
+		// mark all tipsets from the boundary epoch in the Index as reverted as they are not in the current canonical chain
+		log.Infof("Marking tipsets as reverted from height %d", boundaryEpoch)
+		_, err := tx.StmtContext(ctx, si.revertTipsetsFromHeightStmt).ExecContext(ctx, int64(boundaryEpoch))
+		if err != nil {
+			return xerrors.Errorf("error marking tipsets as reverted: %w", err)
+		}
+
+		// Now apply all missing tipsets in reverse order i,e, we apply tipsets in [last matching tipset b/w index and canonical chain,
+		// current chain head]
+		for i := len(tipsetStack) - 1; i >= 0; i-- {
+			curTs := tipsetStack[i]
+			if err := si.indexTipset(ctx, tx, curTs); err != nil {
```

Comment:
What happens if we went too far back and don't have messages at this tipset? we're going to get a fatal here and not be able to reconcile.

We probably need a strategy to figure out what the oldest tipset is that we can _actually_ reconcile, not just _need_ to. Ideally we'd check that on the reverse walk but then we'd end up loading messages unnecessarily.

Could we just `indexTipset` as we're doing the reverse walk instead? Do we need them to go in order?